src/mem/epoch/garbage.rs

```diff
@@ -13,7 +13,94 @@ use std::sync::atomic::Ordering::{Relaxed, Release, Acquire};
 
 use mem::ZerosValid;
 
+/// One item of garbage.
+///
+/// Stores enough information to do a deallocation.
+struct Item {
+    ptr: *mut u8,
+    free: unsafe fn(*mut u8),
+}
+
+/// A single, thread-local bag of garbage.
+pub struct Bag(Vec<Item>);
+
+impl Bag {
+    fn new() -> Bag {
+        Bag(vec![])
+    }
+
+    fn insert<T>(&mut self, elem: *mut T) {
+        let size = mem::size_of::<T>();
+        if size > 0 {
+            self.0.push(Item {
+                ptr: elem as *mut u8,
+                free: free::<T>,
+            })
+        }
+        unsafe fn free<T>(t: *mut u8) {
+            drop(Vec::from_raw_parts(t as *mut T, 0, 1));
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Deallocate all garbage in the bag
+    pub unsafe fn collect(&mut self) {
+        let mut data = mem::replace(&mut self.0, Vec::new());
+        for item in data.iter() {
+            (item.free)(item.ptr);
+        }
+        data.truncate(0);
+        self.0 = data;
+    }
+}
+
+// needed because the bags store raw pointers.
+unsafe impl Send for Bag {}
+unsafe impl Sync for Bag {}
+
+/// A thread-local set of garbage bags.
+// FIXME: switch this to use modular arithmetic and accessors instead
```

Comment:
Let's drop this FIXME, assuming you agree that the named bags are actually clearer here.